Add `.uname` and `.codepoint` to style characters inline

aphillips opened the pull request:
Adds styles recommended by I18N in
https://www.w3.org/TR/international-specs for the inline naming of characters and code points.

\[I18N-ACTION-1257\]

aphillips committed:
Add `.uname` and `.codepoint` to style characters inline

Adds styles recommended by I18N in
https://www.w3.org/TR/international-specs for the inline naming of
characters and code points.

\[I18N-ACTION-1257\]

aphillips committed:
Add `.uname` and `.codepoint` to style characters inline

Adds styles recommended by I18N in
https://www.w3.org/TR/international-specs for the inline naming of
characters and code points.

\[I18N-ACTION-1257\]

r12a commented on src/base.css:
```diff
@@ -1344,6 +1346,19 @@ Possible extra rowspan handling
 		font-weight: bold;
 	}
 
+/** Character examples (from the I18N template) *******************************/
```
I'd suggest the following changes to the code.
````

.codepoint bdi {
   font-size: 120%;
   }
.codepoint .uname {
   font-size: 80%;
   margin-inline: 2px;
   letter-spacing: 0.03em;
   text-transform: uppercase;
   }
````

Additional, document-specific CSS styling may be necessary to style the contents of the bdi, usually based on the lang attribute attached to that element. This would apply font-family, and possibly (if the font has unusual metrics) font-size settings.

aphillips replied:
Done!

aphillips replied:
@r12a are you able to merge this?

aphillips committed:
Address @r12a's comment.

aphillips committed:
Address @r12a's comment.

fantasai reviewed (changes requested):
- Please document any additions to the stylesheet in the README.
- It looks like you're trying to use smallcaps here, please actually use smallcaps.

aphillips commented:
@fantasai noted:

> Please document any additions to the stylesheet in the README.

Thanks for the pointer! I will add that to a revision imminently.

> It looks like you're trying to use smallcaps here, please actually use smallcaps.

Do you mean `font-variant: small-caps`? The problem with that is that the name generally is already uppercase and font-variant small-caps only affects the lowercase letters. (It's certainly the case that the code point notation `U+BEEF` is generally already uppercase). This style has the same _effect_ as small-caps, but on mixed case and uppercase.

Of course, I'm probably missing a feature of CSS I don't know. Happy to change to the right thing.

aphillips commented:
@fantasai:

> Thanks for the pointer! I will add that to a revision imminently.

Hmm. The file `src/README.html` is static HTML. Am I supposed to edit the boilerplate stuff? Should I just add my stuff? Or worry about the "tragedy of the commons"?

fantasai commented:
> Do you mean font-variant: small-caps? The problem with that is that the name generally is already uppercase and font-variant small-caps only affects the lowercase letters.

* Option A: `font-variant: all-small-caps` https://www.w3.org/TR/css-fonts-4/#font-variant-caps-prop
* Option B: `text-transform: lowercase; font-variant: small-caps`

fantasai commented:
> Am I supposed to edit the boilerplate stuff?

Edit whatever needs editing. It's not generated, because this is the template against which the generators are designed. :)

aphillips committed:
Fix .uname to use font-variant

Thanks to @fantasai

aphillips committed:
Fix .uname to use font-variant

Thanks to @fantasai

aphillips commented:
(The readme changes will take me some time. I'll update this PR again when we're ready. The change 244dd86 does ***not*** represent "done")

aphillips commented:
@fantasai We're contemplating further changes to the styling itself. A [question has some up about using color](https://github.com/w3c/i18n-glossary/pull/34#discussion_r1156101131) to set the character name apart from the text--and specifically *which* color to use. Is there any guidance on how to choose a color to match the general color scheme? Could you maybe suggest a color to use? Or should we avoid it?

fantasai commented:
@aphillips I don't have a good sense of colors. :( What part of this were you planning to color? Can you post an example of the markup?

xfq commented:
@fantasai We used teal initially, but since it is the same colour as the link colour in i18n articles, we changed it to brown later.

See https://w3c.github.io/bp-i18n-specdev/#example-example-of-a-character-reference for an example.

r12a commented:
Actually, i think we're good on this now.  We experimented with various alternatives before settling on brown.

aphillips commented:
Thanks @fantasai, @xfq, and @r12a. When I get a chance, I will update this PR with our various modifications, including the updated readme.

fantasai commented:
@r12a Color looks nice afaict :) The rule about not including a space between the character and the U+ code seems weird and counter-intuitive, though. I think it would be better to expect it. The font-size change is also a bit disturbing; could we just make it bold? Maybe something like:

```
.codepoint > bdi {
  font-weight: bold;
}
.codepoint > bdi:hover {
  transform-origin: center;
  scale: 130%;
}
```

Another question I have is whether that outermost `span` should be `code`.

And lastly, shouldn't `<code>` default to `translate=no`?

xfq commented:
> @r12a Color looks nice afaict :) The rule about not including a space between the character and the U+ code seems weird and counter-intuitive, though. I think it would be better to expect it.

I think so too. However, I think using styles to control is for the ease of customization (because it is difficult to change the width of U+0020).

> The font-size change is also a bit disturbing; could we just make it bold? Maybe something like:
> 
> ```css
> .codepoint > bdi {
>   font-weight: bold;
> }
> .codepoint > bdi:hover {
>   transform-origin: center;
>   scale: 130%;
> }
> ```

I think it's to make the character itself (rather than the code point and Unicode name) stand out. I don't have a strong preference here, let's see what others think. I made a comparison document [here](https://xfq.github.io/testing/uname/), so that people can see the rendered document easily.

> Another question I have is whether that outermost `span` should be `code`.

I don't think the outermost `span` should be `code`, because the character is not code.

> And lastly, shouldn't `<code>` default to `translate=no`?

I think code should default to `translate=no`, especially for the non-comment part of the code, but I didn't check if it's specified in the HTML Standard. If the code is translated, it may have an impact on the semantics of the code. Although variable names can be translated technically most of the time, it is not common to do so in practice. If translation is needed, we can add `translate` for those `code` elements.

aphillips commented:
I think we would want to avoid `font-weight: bold` for a variety of reasons, not the least of which is that support for many scripts can be limited in terms of different weights/styles of fonts being available. It can also make certain characters look different or more difficult to read (or turn into tofu). This may not affect other people's specifications very much, but it might prove a bother for us. Plain styles/normal weights are generally available in fallback fonts or are the _most_ available when it comes to complex things such as conjuncts in Indic scripts. 

I do like the idea of using `:hover` for the size.

aphillips committed:
Update the styles and readme
- match the text in internation-specs (bp-i18n-specdev) with local edits
  as needed (mainly to link to specdev)
- update the color to brown
- relax the "no spacing" requirement
- fix a bug in the document's index

aphillips committed:
Update the styles and readme
- match the text in internation-specs (bp-i18n-specdev) with local edits
  as needed (mainly to link to specdev)
- update the color to brown
- relax the "no spacing" requirement
- fix a bug in the document's index

aphillips commented:
I have updated the readme and the stylesheet to match the changes in specdev. Note that I needed to correct an error in the TOC of the readme.

The section about the character template is slightly different from the section in specdev in that it points to specdev (the one in TR) in a couple of places. I also relaxed the wording about not including a space between `bdi` and `span class=uname`. Ready?

fantasai commented on src/base.css:
```diff
@@ -1344,6 +1346,21 @@ Possible extra rowspan handling
+}
+
+.uname {
+    font-family: Arial, monospace;
```
Setting the font family to Arial, monospace seems super weird. What's the justification?

aphillips replied:
We like how Arial looks and know that it is widely available. It is different from the body font of a spec. When Arial is not available, falling back to a sans serif font is not necessarily going to produce the contrast desired. We dithered about just making the font _be_ monospace.

fantasai replied:
Alternatively you could just let the font family inherit? It's defined to be `sans-serif`, which could be anything sans-serif -- including Arial.

You already have color and capitalization shifts here, I don't think a font family shift is also necessary (particularly given this text is reasonably understandable even if there was no formatting change at all!).

aphillips replied:
Agreed. See below for further changes.

fantasai commented on src/base.css:
```diff
@@ -1344,6 +1346,21 @@ Possible extra rowspan handling
+
+.uname {
+    font-family: Arial, monospace;
+	font-size: 85%;
```
Didn't you want `font-variant: all-small-caps` here?

aphillips replied:
At one time we did the text transform, but the consensus was that this would mask bugs in the user's input of the Unicode character name. Instead of making the capital letters a style, we simulated small caps on text we presume to be all caps to start with.

fantasai replied:
It's not quite the same effect: https://www.w3.org/TR/css-fonts-4/#font-variant-caps-prop

> would mask bugs in the user's input of the Unicode character name

Is casing really a likely error? I imagine the most likely error is that it'll be written out in all lowercase or in title caps, both of which wouldn't look terribly wrong either way. And it doesn't seem like it would cause a meaningful loss in information, the uppercasing of these names is basically just a convention, isn't it?

aphillips replied:
This is super-helpful. I made the change, removing the font-family in favor of `all-small-caps`. When I did that I noticed that the digits in a code point (ASCII decimal digits 0-9) display badly compared to small caps for hex digits A-F, so I also added a `font-variant-numeric` with value `oldstyle-nums`. @r12a: have another look at this and see what you think.

fantasai commented on src/base.css:
```diff
@@ -1344,6 +1346,21 @@ Possible extra rowspan handling
+/** Character examples (from the I18N template) *******************************/
+.codepoint bdi {
+    line-height: 1em;
+    font-size: 120%;
```
I don't think the inline font-size change is great. It interferes with understanding the relationship of the glyph to the rest of the text. E.g. for é it makes it look like maybe this is a special lowercase e that is the same height as the capital letters (for some weird Unicode-y reason).

fantasai replied:
Btw, if you want to zoom on :hover, I think the code that works is

```
bdi { display: inline-block; }
bdi:hover { transform-origin: center; scale: 150%; }
```

aphillips replied:
Let me try the :hover again. I think I was missing the inline-block part.

aphillips replied:
we chose not to use the hover and we think the size is a better way to emphasize the character than fooling around with bold (for the reasons we gave before: bold is not always available)

fantasai commented on src/base.css:
```diff
@@ -1344,6 +1346,21 @@ Possible extra rowspan handling
 	}
 
+/** Character examples (from the I18N template) *******************************/
+.codepoint bdi {
```
If this is supposed to handle multi-character sequences, should it be .unicode instead of .codepoint?

aphillips replied:
Good point. Perhaps `.codepoints`?

aphillips replied:
Changing the class name will break a lot of our markup ☹️ so we decided to leave as-is

fantasai replied:
In what way?

aphillips replied:
We have many many documents with the class `codepoint` in it. If we rename the class to `codepoints` we'd need to change the markup to match. Of course we could carry a local style of `codepoint` while doing migration......

Ultimately our goal is to automate this so that people don't type in the template but instead generate it from a more compact markup using features in e.g. Bikeshed or ReSpec.

fantasai commented:
(Filed https://github.com/whatwg/html/issues/9364 fwiw.)

aphillips committed:
Changes to address comments
- Removed static larger sizing of codepoint
- Added :hover on bdi to increase size

aphillips committed:
Changes to address comments
- Removed static larger sizing of codepoint
- Added :hover on bdi to increase size

aphillips committed:
Changes after 2023-06-01 I18N telecon
- restored the styles pre-previous commit
- removed :hover
- changed example text later in the README to use template

aphillips committed:
Changes after 2023-06-01 I18N telecon
- restored the styles pre-previous commit
- removed :hover
- changed example text later in the README to use template

aphillips commented:
We discussed this in the I18N teleconference of 2023-06-01 and decided that using `:hover` and making the inline size consistent with the core text was not the right approach. @r12a noted that we want to emphasize the characters rather than the Unicode names and that the styles proposed have been experimented with fairly extensively.

Note that we're working on automation via ReSpec and Bikeshed so that most spec authors won't use these styles directly. They'll use a much more compact notation.

In the change 0aeea32 I also modified text quoted from a CSS spec to use the character template (the better to demonstrate usage). @fantasai have a look at that and see if these are changes that might go upstream 🙂.

aphillips committed:
Address @fantasai's comments
- remove font change
- add font-variant for all-small-caps
- add font-variant-numeric for old style numbers

The last edit was added because all-small-caps looks odd with "normal"
digits.

aphillips committed:
Address @fantasai's comments
- remove font change
- add font-variant for all-small-caps
- add font-variant-numeric for old style numbers

The last edit was added because all-small-caps looks odd with "normal"
digits.

aphillips commented:
@fantasai @r12a Pinging for another look.

fantasai committed:
Clean up README for unicode codepoints section

fantasai committed:
Add background for codepoint rendering

fantasai commented:
@aphillips Please enable maintainer edit permissions

aphillips commented:
@fantasai It already is enabled?

![image](https://github.com/w3c/tr-design/assets/69082/96a6adcf-b0cf-4f61-84b3-9297a845fa7f)

fantasai commented:
@aphillips OK, I figured it out. :) I've pushed three things here for you to take a look:

1. Rebased on top of the latest origin repo state. (This was a force-push. You'll need to pull again from your own repo to get the new state locally, I hope it doesn't cause too much problems...)
2. Cleaned up the README to follow the style of the rest of the README, fixed indentation, added subheadings, etc.
3. Tweaked the styling a bit to add a background around the BDI, because the font size change is hard to interpret without something to set it apart from the main text.

Take a look and let me know what you think.

> I don't think the outermost `span` should be `code`, because the character is not code.

Should the inner `span` be `code`?

(I do think the `bdi` should be `samp`, but BDI + SAMP seems a bit excessive...)

> `font-variant-numeric: oldstyle-nums;`

What are you trying to do with this? Capital letters + oldstyle numbers is a very weird, and generally unintended, combination...

aphillips commented:
@fantasai Thanks. I'll pull again and we'll see if I survive the merge 😀 

`span class=uname` might be `code class=uname`. The Unicode name is kind of code-ish.

I agree that BDI+SAMP would be excessive. An alternative might be `<samp dir=auto lang=??>` on the theory that every character or short character sequence's base direction is probably pretty indicative. @r12a any thoughts?

The `oldstyle-nums` was something I introduced with the current crop of edits. When we use `all-small-caps`, the ASCII digits remain their non-small-cap larger selves (because digits are unicameral). This looks clunky and somewhat ransom-note-like:
![image](https://github.com/w3c/tr-design/assets/69082/52ace0fd-79b9-4b12-857f-d89d62a37122)

Whereas "oldstyle" numerals are meant to appear with a similar x-height:
![image](https://github.com/w3c/tr-design/assets/69082/90cd49a3-b10d-4d0f-9e33-4c7cbf11466e)

Admittedly this is weird styling, because the name is "code" and old style numerals are.... yeah. But functionally the height matching works, making it more attractive?!?!?

fantasai commented:
@aphillips Worst case, if you don't have anything local, you can delete your local copy of the repo and re-clone from `git@github.com:aphillips/tr-design.git`? :)

Wrt oldstyle nums... that's funny, because I thought all-small-caps was supposed to compensate for that. :) https://www.w3.org/TR/css-fonts-4/#example-06f22977 But maybe not. I should raise this question with Myles Maxfield... I suppose that might be an argument for using text-transform + font-size instead of proper small caps. :/

fantasai commented:
> on the theory that every character or short character sequence's base direction is probably pretty indicative

Well, whether it is nor not, `<bdi>` defaults its `dir` attribute to `auto` anyway, so it's functionally the same either way!

aphillips committed:
Update styles per I18N teleconference.
- Removes borders and backgrounds
- Updates the documentation in the stylesheet
- Removes oldstyle numerals

aphillips commented:
(returning to `oldstyle-nums`) I removed that from the style. It appears that this non-matching of digits in `all-small-caps` is a gap/bug in Chrome (I didn't test extensively, FF does it correctly), so it's probably better to ask for fixes there than go "olde skool" with the digits.

aphillips commented:
@r12a Can you review and ensure I didn't miss anything from telecon here? I'd like to merge this per our discussion.

Also, mentioning: w3c/i18n-actions#15

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+
+<pre>
+&lt;span class="codepoint" translate="no">&lt;bdi lang="??">&#xXXXX;&lt;/bdi> 
+&lt;code class="uname">U+XXXX Unicode_character_name&lt;/span>&lt;/code>
```
The whitespace between the 2 lines needs to be removed.

aphillips replied:
I fixed all of these by going to `code` vs. `pre`. I also changed "Unicode_character_name" to "UNICODE CHARACTER NAME ALL IN CAPS" to make clear what we want by modeling good behavior.

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+&lt;code class="uname">U+XXXX Unicode_character_name&lt;/span>&lt;/code>
+</pre>
+
+<p>For invisible characters (such as control characters) or for whitespace, the character and its surrounding <code translate="no" class="kw">bdi</code> element should be omitted:</p>
```
I think this should say:
> For invisible characters (such as control characters) or for whitespace, the character and its surrounding <code translate="no" class="kw">bdi</code> element can be omitted, or the <code translate="no" class="kw">bdi</code> element can be replaced with an image.

aphillips replied:
okay

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+</pre>
+
+<p>Produces output in the page like this: <span class="codepoint" translate="no"><bdi lang="fr">é</bdi><code class="uname">U+00E9 LATIN SMALL LETTER E WITH ACUTE</code></span>.</p>
+</aside>
```
Needs an example of how to use an image.

aphillips replied:
DONE

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+
+	<div class="xref" style="font-size: 80%; padding: 1em; background-color: #fbf4e0;border-radius: 5px;padding: .25em 1em;"><span class="seealso">See also</span>
+	<ul>
+	   <li>[<a href="https://www.w3.org/International/i18n-activity/guidelines/editing#codepoints">I18N Editing Guidelines</a>].</li>
```
This needs to be updated to show how to use an image, etc.  I will do that, so that you can copy the text, if you wish.

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+<p>Produces output in the page like this: <span class="codepoint" translate="no"><bdi lang="fr">é</bdi><code class="uname">U+00E9 LATIN SMALL LETTER E WITH ACUTE</code></span>.</p>
+</aside>
+
+<p>The <code translate="no" class="kw">bdi</code> element is used to ensure that example characters that are right-to-left do not interfere with the layout of the page. Although the elements are shown on separate lines in this section (for clarity), you do not need to include line breaks or a space between the closing <code translate="no" class="kw">bdi</code> and the following <code translate="n ...
```
Replace

> Although the elements are shown on separate lines in this section (for clarity), you do not need to ..

with

> Do not ...

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+
+<p>The <code translate="no" class="kw">bdi</code> element is used to ensure that example characters that are right-to-left do not interfere with the layout of the page. Although the elements are shown on separate lines in this section (for clarity), you do not need to include line breaks or a space between the closing <code translate="no" class="kw">bdi</code> and the following <code translate="n ...
+
+<p>The <code translate="no" class="kw">lang</code> attribute should be filled in appropriately to get the correct font selection for a given context. Examples in East Asian languages (such as Chinese, Japanese, or Korean) or in the Arabic script can sometimes require greater care in choosing a language tag.</p>
```
It's also necessary to declare the font family and size for the language text in the CSS.

aphillips replied:
Not sure what you mean here. Do you want to change this text or is this about something else?

r12a replied:
I think there needs to be a paragraph to remind people using the markup that they will need to use appropriate styling in their style sheet, where necessary, to correctly render the character in the bdi element in terms of font-family and font-size.

aphillips replied:
Please suggest something?

r12a replied:
Make that an extension of the above para, rather than a separate para.  Perhaps add as the second sentence:

> Your style sheet may need to indicate the appropriate font-family and font-size for characters in a given language.

aphillips replied:
I adapted that thusly:

> In some cases, it might also be necessary for you to adjust the <kbd translate="no">font-family</kbd> and/or <kbd translate="no">font-size</kbd> in your own style sheet for characters or character sequences in a given language.

r12a replied:
I think that text sounds good, but i'm not seeing it in the diff.

r12a commented on src/README.html:
```diff
@@ -925,13 +926,159 @@ <h4 id="inline">
+
+</aside>
+
+<p>Long sequences will sometimes list the code points without including the character names, although the character names should be retained where possible for clarity. An example can be found in <a href="https://www.w3.org/TR/international-specs#char_term_def"><cite>Best Practices for International Specifications</cite></a> in the discussion of the composed "family" emoji: <span class="codepoint ...
```
I suggest to change 'although' to 'but'.

r12a commented on src/base.css:
```diff
@@ -1357,6 +1359,22 @@ Possible extra rowspan handling
+
+	.codepoint bdi {
+		line-height: 1em;
+		font-size: 125%;
```
i'm not clear why this is expressed as a percentage, when the other two values here are in ems.  Why not 1.25em?

r12a commented on src/base.css:
```diff
@@ -1357,6 +1359,22 @@ Possible extra rowspan handling
+	}
+
+	.uname {
+		font-variant: all-small-caps;
```
Why has the small caps reappeared. I thought we banished that long ago, so that content authors would immediately see if they typed the names in lowercase - making the text better suited for copy/paste.

aphillips replied:
We restored small caps instead of using `font-size` per fantasai. Of course, this turns up the small caps numbers bug in Chrome (hence the now removed oldstyle-nums--we just turn a blind eye to Chrome's icky appearance for now). Going back to font-size would fix both.

r12a commented on src/base.css:
```diff
@@ -1357,6 +1359,22 @@ Possible extra rowspan handling
+
+	.uname {
+		font-variant: all-small-caps;
+		font-size: 100%;                    
```
The font size appears to have grown also.

Currently, respec_2022.css has the following:
````
.codepoint bdi {
    line-height: 1em;
    font-size: 1.6em;
    padding-inline: 0.25em;
	}

.uname {
	font-size: 75%;
	letter-spacing:0.03em;
	color: brown;
	}
````

r12a commented on src/base.css:
```diff
@@ -1357,6 +1359,22 @@ Possible extra rowspan handling
+
 /** Outdated warning **********************************************************/
 
 .outdated-spec {
```
Give me a short while to update the editor guidelines, so that you can use the examples here.

r12a commented:
Ok, i updated the text at https://www.w3.org/International/i18n-activity/guidelines/editing#codepoints

I also added some CSS to deal with img tags.
````
.codepoint img {
	height: 2em;
	margin-inline: 0.25em;
        vertical-align: bottom;
	}
````

aphillips committed:
Fixing the line-breaks in the templates
- Switched from `pre` to `code` for the template text to eliminate the
  false line breaks
- Changed "Unicode_character_name" to "UNICODE CHARACTER NAME ALWAYS IN
  CAPS" to help make clear what we want by modeling good behavior :-)

aphillips committed:
Revision to match changes in INTERNATIONAL-SPECS

This includes:
- style changes discussed with @r12a
- addition of an 'images' directory with the image of 00A0
- textual changes to match specdev.

The only wording change is to the reference to an example in specdev,
where the link needed to point to that document rather than internally
in this readme.

aphillips commented:
Okay, something is wrong here. I pushed a new commit to my fork, but it doesn't show in the PR.

https://github.com/w3c/tr-design/compare/gh-pages...aphillips:tr-design:gh-pages?expand=1

aphillips committed:
Resolve issues and address @r12a's comments

aphillips commented:
@r12a This should do it.

- Changed `span` to `code` for the uname
- Added back `font-family: Arial, monospace` for uname
- Removed the merge conflict
- Added the image example. 

Have a look and I'll ensure that spec dev mirrors this.

aphillips committed:
Add info about adjusting the style for a given language.

aphillips commented:
@r12a pinging for approval 😸

r12a commented on src/README.html:
```diff
@@ -956,7 +956,7 @@ <h4 id="unicode-codepoints">
 <p>The <code translate="no" class="kw">bdi</code> element is used to ensure that example characters that are right-to-left do not interfere with the layout of the page. Do not include line breaks or a space between the closing <code translate="no" class="kw">bdi</code> and the following <code translate="no" class="kw">span</code> element; spacing and presentation is controlled by styling.</p>
 
-<p>The <code translate="no" class="kw">lang</code> attribute should be filled in appropriately to get the correct font selection for a given context. Examples in East Asian languages (such as Chinese, Japanese, or Korean) or in the Arabic script can sometimes require greater care in choosing a language tag.</p>
+<p>The <code translate="no" class="kw">lang</code> attribute should be filled in appropriately to get the correct font selection for a given context. Examples in East Asian languages (such as Chinese, Japanese, or Korean) or in the Arabic script can sometimes require greater care in choosing a language tag. In some cases, it might also be necessary for you to adjust the <kbd translate="no">font-f ...
```
I think if might be useful to add "for the content of the bdi element" in there somewhere - just so that people don't style the .codepoint element.

In case it helps, the first para of the editorial guidelines now says:

> Use the following markup to represent a character. For certain languages, it may be necessary to style the bdi element with a font-family and font-size.

aphillips committed:
Address @r12a's comment.

aphillips commented:
@fantasai @r12a I think this is ready to merge? I can't do so myself, either because I don't have an approval from @fantasai or possibly I don't have permission.

r12a commented:
/me could do this, but thinks he should leave this for @fantasai, since it's not his repo.

aphillips commented:
@fantasai Can you approve? Thanks!!

aphillips commented:
@fantasai Pinging to bring to the top of your inbox 😉 

Can we please merge this? I'm blocked on pushing the next version of our best practices for spec authors doc on this.

aphillips committed:
Fix one instance of `span` needing to be `code`

fantasai reviewed (changes requested):
I don't understand what happened to this revision: https://github.com/w3c/tr-design/pull/323/commits/f20c65a3ee55d9c654f5c77fa47aa29aa291f641 It looks like you mostly reverted it, and I don't understand why.

Also, please follow the source-formatting conventions of the document you're editing instead of imposing your own on the sections you edit. (And your indentation is inconsistent anyway.)

aphillips commented:
@fantasai Thanks for the review.

Over the course of the _many_ revisions in this PR, I18N reverted some of the styling suggestions made above (while keeping some others, such as the use of `code` in place of `span`). Notably, we got rid of small caps, old-style numerals, and text transformations to better ensure that the underlying text is copy-ready (because mixed case will show in the page and thus prompt fixes to the underlying text). This is documented by the template text thusly:

```html
<code class="uname">U+XXXX UNICODE_CHARACTER_NAME_ALL_IN_CAPS</code>
```

The text you find here is verbatim from our best practices doc. I can adjust indentation there to match this document if necessary. I think it is important to keep these in sync. That text is staged as [this PR (preview)](https://deploy-preview-113--bp-i18n-specdev.netlify.app/#char_ref_unicode_template).

aphillips committed:
Fix indentation.

aphillips committed:
Fix line-breaking for first paragraph of unicode-codepoints section.

aphillips committed:
Ensure I18N styles use tab rather than space indents

xfq commented:
@deniak? Any comments on this PR?

r12a commented:
In our last CCS<->I18n meeting, Fantasai repeated her preference for a space between the bdi and .uname tags.  Although i had argued that in properly semantic markup it would be better to omit the space, so that the gap could be styled accurately, i can see that people are likely to get this wrong quite often.  Therefore, i don't object to adding the space as part of the canonical markup.

deniak commented:
> @deniak? Any comments on this PR?

The PR looks good to me but I'll defer to @fantasai who's much more knowledgeable than me on these topics

aphillips committed:
Add space between `bdi` and `code`

aphillips commented:
@r12a Fair enough, although I suspect most people will generate the markup using the shorthands we're adding to respec and bikeshed. 

I considered making the space optional by rephrasing the sentence about spaces thusly:

>  Spacing and presentation is controlled by styling, so you do not need to include line breaks or a space between the closing <code translate="no" class="kw">bdi</code> and the following <code translate="no" class="kw">code</code> element, although you can if you wish.

But then decided to just make a clean break. I added spaces and removed the spacing from the style sheet. While typing this I realize that I didn't fix the examples, but will add one more push in a few seconds and then we should hopefully be done?

aphillips committed:
Fix examples to have spaces

Note that this restored the `padding-inline` because some places might
forget the space and need it to avoid the example character and uname
butting up against each other.

fantasai commented:
@aphillips What's the benefit of adding images for invisible characters? Is it really worth the cost to a) creating the images b) tracking the images with the specification text c) complicating the stylesheet and markup patterns and d) disrupting the paragraph line spacing?

fantasai committed:
Unicode codepoint: README clean-up and drop img support

* Revert README to the cleaned up version with subheadings from
    985b81a60adca784e1e5ded266f464e1691fe326
  which seems to have gotten lost in merge commit
    56d969aa84d9966dc89bac1218b865b06d6825b7

* Drop support for img representations for invisible characters
  pending further discussion

fantasai commented:
@aphillips So... I broke something. Can you give me write access to your repo so I can try to fix it? :)

aphillips commented:
@fantasai I invited you to my fork... have at it 👍 

Regarding images, @r12a has been using them for some time and feels that they provide better usability. The images helps make invisibles clearer, at least in some cases. An argument can be made that the sometimes detract from the page and that it is a maintenance headache (especially with regard to echidna configuration), but it is somewhat optional. We might make it clearer that it is up the discretion of the author to use images if they so desire and to spell out the maintenance tradeoff.

@r12a, @xfq any additional thoughts?

fantasai reviewed (approved):
@aphillips OK, I'm going to suggest we tackle that question separately. I want to land what we've got for the basic markup since this is getting to be a really complicated PR.

aphillips commented:
@fantasai 

The diff got super icky on the pushes you made, so I'm not sure what you changed, but as far as I can tell this looks good. Okay to merge?

aphillips commented:
@fantasai Argh. Never mind: you already did merge it. A little blind today.

xfq commented:
> @aphillips What's the benefit of adding images for invisible characters? Is it really worth the cost to a) creating the images b) tracking the images with the specification text c) complicating the stylesheet and markup patterns and d) disrupting the paragraph line spacing?

It is not only useful for invisible characters, but also useful for combining characters. If the font doesn't support display with a dotted circle (`U+25CC DOTTED CIRCLE`), then an image is needed.

xfq commented:
@deniak Since we have merged this, I wonder if there is anything we need to do to ship it to /TR specs?

deniak commented:
> @deniak Since we have merged this, I wonder if there is anything we need to do to ship it to /TR specs?

The TR stylesheet has been updated already. See:
* https://www.w3.org/StyleSheets/TR/2021/base.css
* https://www.w3.org/StyleSheets/TR/2021/README.html

xfq commented:
Great. Thank you all!

r12a commented:
The section at https://www.w3.org/StyleSheets/TR/2021/README.html#unicode-codepoints looks out of date to me.

It says:

> For invisible characters (such as control characters) or for whitespace, the character and its surrounding bdi element should be omitted:

 but the latest guidelines should suggest using a graphic, eg.

> For invisible characters (such as control characters) or for whitespace, use an image instead of the character; or you may also omit the character and its surrounding bdi element.
> `<span class="codepoint" translate="no"><img alt="..." src="..."><span class="uname">U+XXXX UNICODE_CHARACTER_NAME_ALL_IN_CAPS</span></span>`

I haven't checked other bits, but it suggests to me that we probably should to make sure that is the latest copy, throughout.

xfq commented:
That's because @fantasai reverted that part, but it did make tr-design inconsistent with the recommendation of the i18n specdev document, and it's something we should try to avoid.
